fedimint-client/src/db.rs

```diff
@@ -262,3 +278,250 @@ impl_db_lookup!(
+            + Send
+            + 'r,
+    >,
+>;
+
+/// `apply_migrations_client` iterates from the on disk database version for the
+/// client module up to `target_db_version` and executes all of the migrations
+/// that exist in the migrations map, including state machine migrations.
+/// Each migration in the migrations map updates the database to have the
+/// correct on-disk data structures that the code is expecting. The entire
+/// process is atomic, (i.e migration from 0->1 and 1->2 happen atomically).
+/// This function is called before the module is initialized and as long as the
+/// correct migrations are supplied in the migrations map, the module
+/// will be able to read and write from the database successfully.
+pub async fn apply_migrations_client(
+    db: &Database,
+    kind: String,
+    target_db_version: DatabaseVersion,
+    migrations: BTreeMap<DatabaseVersion, ClientMigrationFn>,
+    module_instance_id: ModuleInstanceId,
+    decoders: ModuleDecoderRegistry,
+) -> Result<(), anyhow::Error> {
+    db.ensure_global()?;
+
+    let mut global_dbtx = db.begin_transaction().await;
+    migrate_database_version(
+        &mut global_dbtx.to_ref_nc(),
+        target_db_version,
+        Some(module_instance_id),
+        kind.clone(),
+    )
+    .await?;
+
+    let disk_version = global_dbtx
+        .get_value(&DatabaseVersionKey(module_instance_id))
+        .await;
+
+    let db_version = if let Some(disk_version) = disk_version {
+        let mut current_db_version = disk_version;
+
+        if current_db_version > target_db_version {
+            return Err(anyhow::anyhow!(format!(
+                "On disk database version for module {kind} was higher than the code database version."
+            )));
+        }
+
+        if current_db_version == target_db_version {
+            return Ok(());
+        }
+
+        let mut active_states =
+            get_active_states(&mut global_dbtx.to_ref_nc(), module_instance_id).await;
+        let mut inactive_states =
+            get_inactive_states(&mut global_dbtx.to_ref_nc(), module_instance_id).await;
+
+        while current_db_version < target_db_version {
+            let new_states = if let Some(migration) = migrations.get(&current_db_version) {
+                info!(target: LOG_DB, "Migrating module {kind} current: {current_db_version} target: {target_db_version}");
+
+                migration(
+                    &mut global_dbtx
+                        .to_ref_with_prefix_module_id(module_instance_id)
+                        .into_nc(),
+                    module_instance_id,
+                    active_states.clone(),
+                    inactive_states.clone(),
+                    decoders.clone(),
+                )
+                .await?
+            } else {
+                warn!("Missing client db migration for version {current_db_version}");
+                None
```

Comment:
Should this be a hard error in case we ever remove wayyyy too old migrations?

Reply:
Ah I was actually thinking of a different use case, where if a module developer accidentally increments the `DATABASE_VERSION` too high, in which case there would be intermediate versions with no migrations. I didn't want to explicitly fail in that case. 

I think I'll keep it like this for now, if we do make the decision to cleanup old migrations in the future, we should probably add additional sanity checks and change this to an explicit error.